src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,303 @@
+    * @param mode the clipping mode
+    */
+  def clip5PrimeEndOfRead(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) = {
+    if (!rec.getReadUnmappedFlag && rec.getReadNegativeStrandFlag) clipEndOfRead(rec, numberOfBasesToClip, mode)
+    else clipStartOfRead(rec, numberOfBasesToClip, mode)
+  }
+
+  /**
+    * Ensures that there are at least clipLength bases clipped at the 3' end
+    * of the read, _including_ any existing soft and hard clipping.  Calculates any
+    * additional clipping and delegates to [[clipStartOfAlignment()]].
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to be clipped
+    * @param mode the clipping mode
+    */
+  def clip3PrimeEndOfRead(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) = {
+    if (!rec.getReadUnmappedFlag && rec.getReadNegativeStrandFlag) clipStartOfRead(rec, numberOfBasesToClip, mode)
+    else clipEndOfRead(rec, numberOfBasesToClip, mode)
+  }
+
+  /** Computes the number of bases that are available to be clipped in a mapped SAMRecord. */
+  private def numberOfClippableBases(rec: SAMRecord): Int = {
+    rec.getCigar.filter(e => e.getOperator.isAlignment || e.getOperator == Op.INSERTION).map(_.getLength).sum
+  }
+
+  /**
+    * Private method that is the workhorse of all the clipping methods. _Always_ works at the start of
+    * the Cigar/bases/quals, and _always_ attempts to clip >= numberOfBasesToClip.
+    *
+    * @param originalElems the elements in the cigar string to be clipped
+    * @param numberOfBasesToClip the number of bases to attempt to clip
+    * @param bases the array of bases for the read to be clipped
+    * @param quals the array of quality scores for the read to be clipped
+    * @param mode  the clipping mode
+    * @return a tuple of (newCigarElems, readBasesClipped, refBasesClipped, newBases, newQuals)
+    */
+  private def clip(originalElems: Seq[CigarElement],
+                   numberOfBasesToClip: Int,
+                   bases: Array[Byte],
+                   quals: Array[Byte],
+                   mode: ClippingMode): (Seq[CigarElement], Int, Int, Array[Byte], Array[Byte]) = {
+
+    if (originalElems.exists(_.getOperator == Op.PADDING)) throw new IllegalArgumentException("Can't handle reads with padding.")
+
+    val existingHardClip = originalElems.takeWhile(_.getOperator == Op.HARD_CLIP).map(_.getLength).sum
+    val existingSoftClip = originalElems.dropWhile(_.getOperator == Op.HARD_CLIP).takeWhile(_.getOperator == Op.SOFT_CLIP).map(_.getLength).sum
+    val postClipElems = originalElems.dropWhile(e => e.getOperator == Op.HARD_CLIP || e.getOperator == Op.SOFT_CLIP).iterator.bufferBetter
+    var readBasesClipped = 0
+    var refBasesClipped  = 0
+    val newElems = ListBuffer[CigarElement]()
+
+    // At the end of the loop newElems is either:
+    //   a) Empty
+    //   b) Contains a single element which is the remainder of an element that had to be split
```

Comment:
Is this true? A cigar of `====` or a cigar of `10M1D10M` (with 15 clipping) would have `newElems.length > 1` right?